### PR TITLE
Modifies timers API to select autostart state

### PR DESCRIPTION
**Pull request by Voldivh**
Signed-off-by: Voldivh <eloyabmfcv@gmail.com>

This PR addresses the issue #1980. Basically it gives the timers API a variable that enables or disables the start of the timer at initialization. The variable `autostart` is defaulted to `True` in order to avoid any breaks in current use of the API. In order to start the timer if `autostart = false` you can simply call the method `timer->reset()`. This PR is dependant on modifications on this [PR at the RCL package](https://github.com/ros2/rcl/pull/1004) and should be merged after that one.

### CODE REVIEW

File: rclcpp/include/rclcpp/timer.hpp [hunk 1]
```diff
@@ -152,8 +154,11 @@ class TimerBase
 protected:
   Clock::SharedPtr clock_;
   std::shared_ptr<rcl_timer_t> timer_handle_;
+  rclcpp::Context::SharedPtr context_;
+  std::chrono::nanoseconds period_;
 
   std::atomic<bool> in_use_by_wait_set_{false};
+  bool autostart_{false};
```
Comment (alsora) [inline rclcpp/include/rclcpp/timer.hpp:161]: why adding these 3 member variables?
where are they used?
moreover `autostart_` is used only in the constructor, so no need to cache it

Comment (Voldivh) [reply rclcpp/include/rclcpp/timer.hpp]: Sorry, the variables were part of another implementation and forgot to delete them.

File: rclcpp/include/rclcpp/create_timer.hpp [hunk 2]
```diff
@@ -92,7 +96,8 @@ create_wall_timer(
   CallbackT callback,
   rclcpp::CallbackGroup::SharedPtr group,
   node_interfaces::NodeBaseInterface * node_base,
-  node_interfaces::NodeTimersInterface * node_timers)
+  node_interfaces::NodeTimersInterface * node_timers,
+  bool autostart = true)
```
Comment (alsora) [inline rclcpp/include/rclcpp/create_timer.hpp:100]: the new `autostart` argument should added to the doxygen description of this function

File: rclcpp/test/rclcpp/test_timer.cpp [hunk 3]
```diff
@@ -283,3 +298,12 @@ TEST_F(TestTimer, test_failures_with_exceptions)
...
   }
 }
+
+/// Simple test of a timer without autostart
+TEST_F(TestTimer, test_timer_without_autostart)
+{
+  EXPECT_TRUE(timer_without_autostart->is_canceled());
+  // Reset to change start timer
+  timer_without_autostart->reset();
+  EXPECT_FALSE(timer_without_autostart->is_canceled());
```
Comment (alsora) [inline rclcpp/test/rclcpp/test_timer.cpp:308]: we should also check that this timer will now trigger

File: rclcpp/include/rclcpp/create_timer.hpp (hunk 2 above)
Comment (Voldivh) [reply rclcpp/include/rclcpp/create_timer.hpp:101]: Sure thing!

File: rclcpp/test/rclcpp/test_timer.cpp (hunk 3 above)
Comment (Voldivh) [reply rclcpp/test/rclcpp/test_timer.cpp:314]: Done.

**Comment by alsora** [comment]
Looks good to me. 
Full CI including your rcl PR
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17204/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11747/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17630/)

File: rclcpp/test/rclcpp/test_create_timer.cpp [hunk 4]
```diff
@@ -193,3 +193,29 @@ TEST(TestCreateTimer, timer_function_pointer)
...
+    node->get_clock(),
+    rclcpp::Duration(0ms),
+    []() {},
+    nullptr,
+    false);
+
+  EXPECT_TRUE(timer->is_canceled());
+  EXPECT_EQ(timer->time_until_trigger().count(), std::chrono::nanoseconds::max().count());
+
+  timer->reset();
```
Comment (fujitatomoya) [inline rclcpp/test/rclcpp/test_create_timer.cpp:214]: I believe this needs to be well-described in doc section. user application must call `reset` the timer for starting.

besides, i would like to have feedback on https://github.com/ros2/rcl/pull/1004#pullrequestreview-1111587769

CC: @alsora @gbiggs

Comment (fujitatomoya) [reply rclcpp/test/rclcpp/test_create_timer.cpp:214]: I think this still needs to be addressed in doc section for user explanation.

**Comment by emersonknapp** [comment]
Pulls: ros2/rcl#1004, ros2/rclcpp#2005, ros2/system_tests#508, ros2/rclpy#999
Gist: https://gist.githubusercontent.com/emersonknapp/f9353e435d78a1ff796a38b74eaf0547/raw/2a9e6c7645332dcc2c4a6d0fb6f7b66e6b428bb7/ros2.repos
BUILD args:  --packages-above-and-dependencies rcl rcl_action rclcpp rclpy test_rclcpp
TEST args:  --packages-above rcl rcl_action rclcpp rclpy test_rclcpp
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/12240
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/18956/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/13489/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/19709/)

**Comment by fujitatomoya** [comment]
CI is green, see https://github.com/ros2/rcl/pull/1004#issuecomment-1593338990

File: rclcpp/test/rclcpp/test_create_timer.cpp (hunk 4 above)
Comment (fujitatomoya) [reply rclcpp/test/rclcpp/test_create_timer.cpp:214]: @Voldivh can you check this comment?

Comment (Voldivh) [reply rclcpp/test/rclcpp/test_create_timer.cpp:214]: Sure thing! Just to be clear, we are talking about adding more explanation on the .hpp file on the timer class section?

File: rclcpp/include/rclcpp/timer.hpp [hunk 5]
```diff
@@ -53,12 +53,17 @@ class TimerBase
    * \param clock A clock to use for time and sleeping
    * \param period The interval at which the timer fires
    * \param context node context
+   * \param autostart timer state on initialization
+   *
+   * In order to activate a timer that is not started on initialization,
+   * user should call the reset() method.
```
Comment (Voldivh) [inline rclcpp/include/rclcpp/timer.hpp:59]: @fujitatomoya this should address the other comment, could you take a look and let me know?

**Review by fujitatomoya — APPROVED** [review]

**Review by clalancette — APPROVED** [review]

**Pull request merged**